### PR TITLE
Add DataBox

### PR DESCRIPTION
**Pull request by nilsdeppe**

### CODE REVIEW

File: src/DataStructures/DataBoxTag.hpp [hunk 1]
```diff
@@ -0,0 +1,334 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp]: add the **best** words

File: src/DataStructures/DataBoxTag.hpp [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+ * \derivedrequires
+ * - type alias `type` of the type this DataBoxTag represents
+ * - `static constexpr DataBoxString_t` that is the same as the type name
+ * \see DataBox DataBoxPrefix DataBoxString_t get_tag_name
+ */
+struct DataBoxTag {};
+
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Functional Tags for the DataBox inherit from this type
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp]: Functional -> Compute item

Comment (nilsdeppe) [reply src/DataStructures/DataBoxTag.hpp]: Move to PrefixTag

File: src/DataStructures/DataBoxTag.hpp [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+ * \ingroup DataBoxGroup
+ * \brief Tags for the DataBox inherit from this type
+ *
+ * \details
+ * Used to mark a type as being a DataBoxTag so that it can be used in a
+ * DataBox.
+ *
+ * \derivedrequires
+ * - type alias `type` of the type this DataBoxTag represents
+ * - `static constexpr DataBoxString_t` that is the same as the type name
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp:47]: Make sure this is called `label`

File: src/DataStructures/DataBoxTag.hpp [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+ * \see DataBox DataBoxPrefix DataBoxString_t get_tag_name
+ */
+struct DataBoxTag {};
+
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Functional Tags for the DataBox inherit from this type
+ *
+ * \details
+ * Used to mark a type as being a DataBoxTag where the label is a prefix to the
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp]: `label` instead of label

Comment (nilsdeppe) [reply src/DataStructures/DataBoxTag.hpp]: Add a code example of what these look like

File: src/DataStructures/DataBoxTag.hpp [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+ * DataBoxTag that is a member type alias `tag`.
+ *
+ * \derivedrequires
+ * - type alias `tag` of the DataBoxTag that the function operated on
+ * \see DataBox DataBoxTag DataBoxString_t get_tag_name
+ */
+struct DataBoxPrefix : DataBoxTag {};
+
+/*! \ingroup DataBoxGroup
+ * \brief Marks a DataBoxTag as being a compute item that executes a function
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp:87]: Add code example and describe what needs to be inside

File: src/DataStructures/DataBoxTag.hpp [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+namespace detail {
+// @{
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Check if a Tag has a label
+ *
+ * \details
+ * Check if a type `T` has a static member variable named `label`.
+ *
+ * \usage
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp:120]: keep usage, remove semantics

File: src/DataStructures/DataBoxTag.hpp [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+ * For any type `T`
+ * \code
+ * using result = db::TagHasLabel<T>;
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp]: make test for this

File: src/DataStructures/DataBoxTag.hpp [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Get the name of a DataBoxTag, including prefixes
+ *
+ * \details
+ * Given a DataBoxTag returns the name of the DataBoxTag as a std::string. If
+ * the DataBoxTag is also a DataBoxPrefix then the prefix is added.
+ *
+ * \tparam Tag the DataBoxTag whose name to get
+ * \return string hold the DataBoxTag's name
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp]: string holding

File: src/DataStructures/DataBoxTag.hpp [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+ * \brief Compute a hash of the `label` of a DataBoxTag
+ *
+ * \details
+ * Given a DataBoxTag returns a `value` of type `value_type` which is the hash
+ * of the label of the DataBoxTag, including any prefix. The hashing algorithm
+ * is based on what is used for std::typeinfo in libcxx v3.9.0.
+ *
+ * \tparam Tag the DataBoxTag whose name to get
+ * \metareturns
+ * value membor of type value_type representing the hash of the DataBoxTag
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxTag.hpp]: member

File: tests/Unit/DataStructures/Test_DataBox.cpp [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+  static constexpr db::DataBoxString_t label = "ComputeTag1";
+  static constexpr auto function = append_word;
+  using argument_tags = typelist<Tag2, ComputeTag0>;
+};
+
+struct TagTensor : db::ComputeItemTag {
+  static constexpr db::DataBoxString_t label = "TagTensor";
+  static constexpr auto function = get_tensor;
+  using argument_tags = typelist<>;
+};
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataBox.cpp:60]: Add tests for prefix tags

File: tests/Unit/DataStructures/Test_DataBox.cpp [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+TEST_CASE("Unit.DataStructures.DataBox.get_tag_from_box",
+          "[Unit][DataStructures]") {
+  auto original_box =
+      db::create<db::AddTags<test_databox_tags::Tag0, test_databox_tags::Tag1,
+                             test_databox_tags::Tag2>,
+                 db::AddComputeItemsTags<test_databox_tags::ComputeTag0,
+                                         test_databox_tags::ComputeTag1>>(
+          3.14, std::vector<double>{8.7, 93.2, 84.7}, "My Sample String"s);
+  const std::string compute_string =
+      db::get_tag_from_box<std::string>(original_box, "ComputeTag1");
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataBox.cpp]: `get_item_from_box`

File: tests/Unit/DataStructures/Test_DataBox.cpp [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+// [[OutputRegex, Could not find the tag named "TagTensor__" in the DataBox]]
+TEST_CASE("Unit.DataStructures.DataBox.HelpersBadTensorFromBox",
+          "[Unit][DataStructures]") {
+  auto original_box =
+      db::create<db::AddTags<test_databox_tags::Tag0, test_databox_tags::Tag1,
+                             test_databox_tags::Tag2>,
+                 db::AddComputeItemsTags<test_databox_tags::TagTensor>>(
+          3.14, std::vector<double>{8.7, 93.2, 84.7}, "My Sample String"s);
+
+  auto tag_tensor = get_tensor_from_box(original_box, "TagTensor__");
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataBox.cpp]: cast to void

Comment (nilsdeppe) [reply tests/Unit/DataStructures/Test_DataBox.cpp]: make return type: `std::pair<std::vector<std::string>, std::vector<DataVector>>`

File: src/DataStructures/DataBox.hpp [hunk 13]
```diff
@@ -0,0 +1,967 @@
...
+template <typename DependencyGraph, typename Vertices>
+struct ResetComputeItems;
+}  // namespace detail
+
+// @{
+/*!
+ * \ingroup TypeTraits DataBox
+ * \brief Determines if a type `T` is as db::DataBox
+ *
+ * \effects Inherits form std::true_type if `T` is a specialization of
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: from

File: src/DataStructures/DataBox.hpp [hunk 14]
```diff
@@ -0,0 +1,967 @@
...
+      std::is_same<void, Caller>::value, sub_tree,
+      tmpl::push_back<sub_tree, tmpl::edge<Callee, Caller>>>;
+};
+}  // namespace detail
+
+/*!
+ * \ingroup DataBoxGroup
+ * \brief A DataBox stores objects that can be retrieved by using Tags
+ * \warning
+ * The order of the tags in DataBoxes returned by functions depends on
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: by `create` and `create_from`

File: src/DataStructures/DataBox.hpp [hunk 15]
```diff
@@ -0,0 +1,967 @@
...
+   * true
+   *
+   * \return A DataBox with items described by Tags in `AddTags` and
+   * values `args...`, and compute items described by Tags in `AddComputeItems`
+   */
+  template <typename AddTags, typename AddComputeItems, typename... Args>
+  static constexpr auto create(Args&&... args);
+
+  /*!
+   * \brief Helper function called by db::create to call the constructor
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: `db::create_from`

File: src/DataStructures/DataBox.hpp [hunk 16]
```diff
@@ -0,0 +1,967 @@
...
+   * \requires Type `T` is one of the Tags corresponding to an object stored in
+   * the DataBox
+   *
+   * `mutate()` is similar to get, however it allows altering the value of the
+   * item in the DataBox.
+   * \return The object corresponding to the Tag `T`
+   */
+  template <typename T>
+  constexpr item_type<T>& mutate() noexcept {
+    return data_.template get<T>().mutate();
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: Add static_assert that `T` is not a compute item

File: src/DataStructures/DataBox.hpp [hunk 17]
```diff
@@ -0,0 +1,967 @@
...
+                              full_compute_items{},
+                              std::forward<Args>(args)...);
+}
+
+template <template <typename...> class TagsLs, typename... Tags>
+template <typename RemoveTags, typename AddTags, typename AddComputeItems,
+          typename Box, typename... Args>
+constexpr auto DataBox<TagsLs<Tags...>>::create_from(const Box& box,
+                                                     Args&&... args) {
+  static_assert(tt::is_a<::db::DataBox, Box>::value,
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: use variable templates

File: src/DataStructures/DataBox.hpp [hunk 18]
```diff
@@ -0,0 +1,967 @@
...
+    data.template get<T>() =
+        make_deferred([lazy_function = data.template get<VariablesTag>()]() {
+          return lazy_function.get().template get<T>();
+        });
+    add_variables_compute_tags_to_box<VariablesTag>(data, typelist<Ts...>{});
+  }
+
+  // add_variables_tags_to_box is used to add the Tensor's inside a Variables
+  // object into the DataBox. If the Tag does not hold a Variables then we
+  // have nothing to add, otherwise we recurse through the Variables we need
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: Variables to Tensors

File: src/DataStructures/DataBox.hpp [hunk 19]
```diff
@@ -0,0 +1,967 @@
...
+ *  \details
+ *  Creates a new DataBox holding types Tags::type filled with the arguments
+ *  passed to the function. Compute items must be added so that the dependencies
+ *  of a compute item are added before the compute item. For example, say you
+ *  have compute items `A` and `B` where `B` depends on `A`, then you must
+ *  add them using `db::AddComputeItemsTags<A, B>`.
+ *
+ *  \example
+ *  \snippet Test_DataBox.cpp create_databox
+ *  \see new_data_box
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: update this name to 'create_from`

File: src/DataStructures/DataBox.hpp [hunk 20]
```diff
@@ -0,0 +1,967 @@
...
+ *  \tparam AddTags the tags of the args being added
+ *  \param args the data to be added to the DataBox
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: `\tparam AddComputeItems`

File: src/DataStructures/DataBox.hpp [hunk 21]
```diff
@@ -0,0 +1,967 @@
...
+ * \ingroup DataBoxGroup
+ * \brief Create a new DataBox from an existing one adding or removing items
+ * and compute items
+ *
+ * \example
+ * Removing an item or compute item is done using:
+ * \snippet Test_DataBox.cpp create_from_remove
+ * Adding an item is done using:
+ * \snippet Test_DataBox.cpp create_from_add_item
+ * Adding an compute item is done using:
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: an to `a`

File: src/DataStructures/DataBox.hpp [hunk 22]
```diff
@@ -0,0 +1,967 @@
...
+ * \snippet Test_DataBox.cpp create_from_add_item
+ *
+ * \see create DataBox get_tags_from_box
+ * \tparam RemoveTags typelist of Tags to remove
+ * \tparam AddTags typelist of Tags corresponding to the arguments to be added
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: `\tparam AddComputeItems`

File: src/DataStructures/DataBox.hpp [hunk 23]
```diff
@@ -0,0 +1,967 @@
...
+ * \param box the DataBox the new box should be based off
+ * \param args the items or compute items to add to the DataBox
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: remove "compute items" since that's just a FALSE FACT!

File: src/DataStructures/DataBox.hpp [hunk 24]
```diff
@@ -0,0 +1,967 @@
...
+constexpr const Type& get_tag_from_box(const DataBox<TagLs>& box,
+                                       const std::string& tag_name) {
+  using Tag = tmpl::front<Tags>;
+  return get_tag_name<Tag>() == tag_name
+             ? box.template get<Tag>()
+             : get_tag_from_box<Type, tmpl::pop_front<Tags>>(box, tag_name);
+}
+}  // namespace detail
+
+/// \ingroup DataBoxGroup
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: write a less terrible doxygen comment

File: src/DataStructures/DataBox.hpp [hunk 25]
```diff
@@ -0,0 +1,967 @@
...
+ *  \semantics
+ *  For tags `Tags...` in a DataBox `box`, and a function `func` that takes
+ *  `sizeof...(Tags)` arguments of types `typename Tags::type...`,  and
+ *  `sizeof...(Args)` arguments of types `Args...`,
+ *  \code
+ *  result = func(box, box.get<Tags>()..., args...);
+ *  \endcode
+ *
+ *  \example
+ *  TODO(): make example
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: Do this for real...

File: src/DataStructures/DataBox.hpp [hunk 26]
```diff
@@ -0,0 +1,967 @@
...
+  return detail::Apply<TagsLs>::apply(f, box, std::forward<Args>(args)...);
+}
+
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Apply the function `f` with argument Tags `TagLs` from DataBox `box`
+ *  and `box` as the first argument
+ *
+ *  \details
+ *  Apply the function `f` with arugments that are of type `Tags::type` where
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: arugments to arguments

File: src/DataStructures/DataBox.hpp [hunk 27]
```diff
@@ -0,0 +1,967 @@
...
+  }
+};
+}  // namespace detail
+
+/*!
+ *  \ingroup DataBoxGroup
+ *  \brief Apply the function `f` with argument Tags `TagLs` from DataBox `box`
+ *
+ *  \details
+ *  Apply the function `f` with arugments that are of type `Tags::type` where
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: arugments to arguments

File: src/DataStructures/DataBox.hpp [hunk 28]
```diff
@@ -0,0 +1,967 @@
...
+ *  \endcode
+ *  \return `decltype(func(box, box.get<Tag1>(), box.get<Tag2>(), a1, a2))`
+ *
+ *  \semantics
+ *  For tags `Tags...` in a DataBox `box`, and a function `func` that takes as
+ *  its first argument a value of type`decltype(box)`,
+ *  `sizeof...(Tags)` arguments of types `typename Tags::type...`, and
+ *  `sizeof...(Args)` arguments of types `Args...`,
+ *  \code
+ *  result = func((box, box.get<Tags>()..., args...);
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: remove extra `(`

File: src/DataStructures/DataBox.hpp [hunk 29]
```diff
@@ -0,0 +1,967 @@
...
+ *
+ *  \example
+ *  TODO(): make example
+ *  \see apply DataBox
+ *  \tparam TagsLs typelist of Tags in the order that they are to be passed to
+ *  `f`
+ *  \param f the function to apply
+ *  \param box the DataBox out of which to retrieve the Tags and to pass to `f`
+ *  \param args the arguments to pass to the function that are not in the
+ *  DataBox, box
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: `box`

File: src/DataStructures/DataBox.hpp [hunk 30]
```diff
@@ -0,0 +1,967 @@
...
+ *
+ *  \example
+ *  TODO(): make example
+ *  \see apply_with_box DataBox
+ *  \tparam TagsLs typelist of Tags in the order that they are to be passed to
+ *  `f`
+ *  \param f the function to apply
+ *  \param box the DataBox out of which to retrieve the Tags and to pass to `f`
+ *  \param args the arguments to pass to the function that are not in the
+ *  DataBox, box
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: `box`

File: src/DataStructures/DataBoxHelpers.hpp [hunk 31]
```diff
@@ -0,0 +1,134 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class
```
Comment (nilsdeppe) [inline src/DataStructures/DataBoxHelpers.hpp]: write better words

File: src/DataStructures/DataBoxTag.hpp [hunk 32]
```diff
@@ -44,26 +45,58 @@ using DataBoxString_t = const char* const;
...
 /*! \ingroup DataBoxGroup
  * \brief Marks a DataBoxTag as being a compute item that executes a function
+ *
+ * \details
+ * A compute item tag wraps a callable such as a function, lambda, or an object
+ * with a call operator. The callable is a member variable with the name
+ * `function`. The compute item tag must also have a label, same as the DataBox
+ * tag. A type alias `argument_tags` that is a typelist of the tags that will
+ * be retrieved from the DataBox and whose data will be passed to the callable.
+ *
```
Comment (osheamonn) [inline src/DataStructures/DataBoxTag.hpp]: It'd be worth mentioning that compute items tags wraps something which along with being callable must be stateless also.

**Review by osheamonn — APPROVED** [review]

**Review by osheamonn — APPROVED** [review]

**Review by osheamonn — APPROVED** [review]

**Pull request merged**